shell/src/main/java/alluxio/cli/fs/command/CheckConsistencyCommand.java

```diff
@@ -122,43 +151,85 @@ private void runConsistencyCheck(AlluxioURI path, boolean repairConsistency) thr
+      CompletionService<Boolean> completionService = new ExecutorCompletionService<>(svc);
+      ConcurrentHashSet<Exception> exceptions = new ConcurrentHashSet<>();
+
+      int totalUris = inconsistentUris.size();
       for (AlluxioURI inconsistentUri : inconsistentUris) {
-        URIStatus status = mFileSystem.getStatus(inconsistentUri);
-        if (status.isFolder()) {
-          inconsistentDirs.add(inconsistentUri);
-          continue;
-        }
-        System.out.println("repairing path: " + inconsistentUri);
-        DeletePOptions deleteOptions = DeletePOptions.newBuilder().setAlluxioOnly(true).build();
-        mFileSystem.delete(inconsistentUri, deleteOptions);
-        mFileSystem.exists(inconsistentUri);
-        System.out.println(inconsistentUri + " repaired");
-        System.out.println();
+        completionService.submit(() -> {
+          try {
+            URIStatus status = mFileSystem.getStatus(inconsistentUri);
+            if (status.isFolder()) {
+              inconsistentDirs.add(inconsistentUri);
+              return;
+            }
+            System.out.println("repairing path: " + inconsistentUri);
+            DeletePOptions deleteOptions = DeletePOptions.newBuilder().setAlluxioOnly(true).build();
+            mFileSystem.delete(inconsistentUri, deleteOptions);
+            mFileSystem.exists(inconsistentUri);
+            System.out.println(inconsistentUri + " repaired");
+            System.out.println();
+          } catch (AlluxioException | IOException e) {
+            exceptions.add(e);
+          }
+        }, true);
       }
+
+      waitForTasks(completionService, totalUris, exceptions);
+
+      int totalDirs = inconsistentDirs.size();
       for (AlluxioURI uri : inconsistentDirs) {
-        DeletePOptions deleteOptions =
-            DeletePOptions.newBuilder().setAlluxioOnly(true).setRecursive(true).build();
-        System.out.println("repairing path: " + uri);
-        mFileSystem.delete(uri, deleteOptions);
-        mFileSystem.exists(uri);
-        System.out.println(uri + "repaired");
-        System.out.println();
+        completionService.submit(() -> {
+          try {
+            DeletePOptions deleteOptions =
+                DeletePOptions.newBuilder().setAlluxioOnly(true).setRecursive(true).build();
+            System.out.println("repairing path: " + uri);
+            mFileSystem.delete(uri, deleteOptions);
+            mFileSystem.exists(uri);
+            System.out.println(uri + "repaired");
+            System.out.println();
+          } catch (AlluxioException | IOException e) {
+            exceptions.add(e);
+          }
+        }, true);
       }
+      waitForTasks(completionService, totalDirs, exceptions);
+      svc.shutdown();
+    }
+  }
+
+  private void waitForTasks(CompletionService svc, int nTasks, Collection<Exception> exceptions)
+      throws IOException {
+    for (int i = 0; i < nTasks; i++) {
+      try {
+        svc.take();
```

Comment:
this could wait forever?

Reply:
It's possible that it hangs forever, but it will only hang until there is a completed task available in the queue (regardless of whether the task finished successfully or with an error). That is, it would only hang if a Runnable from the service hung forever - if that's the case the repair portion of the command would have hung forever before regardless if this was parallelized or not